gravitee-apim-gateway/gravitee-apim-gateway-core/src/main/java/io/gravitee/gateway/jupiter/core/v4/endpoint/DefaultEndpointManager.java

```diff
@@ -0,0 +1,177 @@
+import io.gravitee.definition.model.v4.Api;
+import io.gravitee.definition.model.v4.endpointgroup.Endpoint;
+import io.gravitee.definition.model.v4.endpointgroup.EndpointGroup;
+import io.gravitee.gateway.jupiter.api.connector.endpoint.EndpointConnector;
+import io.gravitee.gateway.jupiter.api.context.DeploymentContext;
+import io.gravitee.plugin.endpoint.EndpointConnectorPluginManager;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class DefaultEndpointManager extends AbstractService<EndpointManager> implements EndpointManager {
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultEndpointManager.class);
+
+    private final Api api;
+    private final EndpointConnectorPluginManager endpointConnectorPluginManager;
+    private final DeploymentContext deploymentContext;
+    private ManagedEndpointGroup defaultGroup;
+    private final Map<String, ManagedEndpointGroup> groupsByName;
+    private final Map<String, ManagedEndpoint> endpointsByName;
+
+    public DefaultEndpointManager(
+        final Api api,
+        final EndpointConnectorPluginManager endpointConnectorPluginManager,
+        final DeploymentContext deploymentContext
+    ) {
+        this.api = api;
+        this.endpointsByName = new ConcurrentHashMap<>(1);
+        this.groupsByName = new ConcurrentHashMap<>(1);
+        this.endpointConnectorPluginManager = endpointConnectorPluginManager;
+        this.deploymentContext = deploymentContext;
+    }
+
+    @Override
+    public ManagedEndpoint next() {
+        return next(EndpointCriteria.ENDPOINT_UP);
+    }
+
+    @Override
+    public ManagedEndpoint next(final EndpointCriteria criteria) {
+        final String name = criteria.getName();
+
+        if (name == null) {
+            if (defaultGroup != null && criteria.matches(defaultGroup)) {
+                return defaultGroup.next();
+            }
+        } else {
+            // First try to find an endpoint by name.
+            final ManagedEndpoint managedEndpoint = endpointsByName.get(name);
+
+            if (managedEndpoint != null) {
+                if (criteria.matches(managedEndpoint)) {
+                    return managedEndpoint;
+                }
+                return null;
+            }
+
+            final ManagedEndpointGroup managedGroup = groupsByName.get(name);
+
+            if (managedGroup != null && criteria.matches(managedGroup)) {
+                return managedGroup.next();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        for (EndpointGroup endpointGroup : api.getEndpointGroups()) {
+            final ManagedEndpointGroup managedEndpointGroup = createAndStartGroup(endpointGroup);
+
+            if (defaultGroup == null) {
```

Comment:
Just a though: Do we still want to base the default group on the first one ? Shouldn't we offer something more explicit on the endpointgroup configuration ?

Reply:
I don't expect having more than one endpoint group very often but that could be something to introduce later if it makes sense 👍